text/0000-anonymous-variants.md

```diff
@@ -0,0 +1,306 @@
+
+However, with these extras comes complexity. RFCs of this kind generally describe their proposed types with extra features which would supply tangible improvements to Rust should they be implemented. However, none of these RFCs have actually been approved for implementation, and arguments against their approval include the complexity associated with these extra features as well as possible ambiguities and undesired interactions of the type, features and all, with the rest of the language. The general idea of anonymous enum-like types people can seem to get behind, but the extras are what get most proposals of this kind. 
+
+This RFC differs from other RFCs in the same vein by deemphasizing ergonomic features, instead focusing on a simple base instantiation of the general idea of anonymous sum types which can be implemented and processed with relative ease, and which library writers and future RFCs can build on top of with relative ease. The RFC itself may be long, but most of this extra length is not used for more detail, but to explain the decisions behind the RFC and how it may be executed and utilized. 
+
+Not to say that this feature won't be useful by itself. Even as described and without ecosystem extras, this feature can still be used reasonably nicely. Here it is used to combine the possible error types of a function with less boilerplate than a single-shot error enum, a commonly cited potential use of ad-hoc sum and algebraic union types: 
+```
+use std::rc::Weak;
+use std::option::NoneError;
+use std::num::ParseIntError;
+
+fn multiple_errors(val: Weak<str>) -> Result<i64, (NoneError|ParseIntError)> {
+
+    // If None, a Result error holding an anonymous variant for NoneError is
+    // returned early
+    let strref = Weak::upgrade(val).ok_or_else(|| _::0(NoneError))? 
+    
+    // If Err, a Result error holding an anonymous variant for ParseIntError
+    // is returned
+    let num = i64::from_str_radix(strref, 10_u32).or_else(_::1)?
+    
+    Result::Ok(num)
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Concise guide
+------
+Anonymous variant types are anonymous sum types, which mean, like enums, their possible values are several different variants. They are declared like tuples, but with vertical bars instead of commas. The variants are referred to for construction and matching by the anonymous variant type name or a placeholder for one, followed by two colons, followed by a (zero-indexed) number for the variant. Like enums, variant order in an anonymous variant type matters, and variants are not automatically combined together or rearranged. 
+```
+// Declare an anonymous variant type
+let x = (i32 | &str)::0(1_i32);
+
+// And then match on it
+match x {
+    (_ | _)::0(val) => assert_eq!(val, 1_i32),
```

Comment:
nit: `(_) =>` is already stabilized as a [valid](https://play.rust-lang.org/?version=nightly&gist=cc2d11c8b9f2aa878e6deb00269f7201) (though warned-about) pattern in nightly, so I suspect this needs to be `<(_ | _)>::0(val) =>` for the same reason you [can't](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=c41e12c123d1073b846404b81c4a4613) do `()::clone(&())`.

Reply:
Changed, especially since the angled brackets are also required for tuples. I'm not particularly fond of the kirby-boss syntax, but consistency helps ease implementation, which is my primary concern.

Reply:
Changed back, it's now an unresolved question, with me leaning towards the syntax depicted above.